// @benjstephenson/kittens-ts/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ord = exports.Eq = exports.m = exports.Sg = exports.TE = exports.T = exports.V = exports.internal = exports.NEA = exports.functions = exports.O = exports.E = exports.A = void 0;
exports.A = __importStar(require("./Array"));
exports.E = __importStar(require("./Either"));
exports.O = __importStar(require("./Option"));
exports.functions = __importStar(require("./functions"));
exports.NEA = __importStar(require("./NonEmptyArray"));
exports.internal = __importStar(require("./hkt"));
exports.V = __importStar(require("./Validation"));
exports.T = __importStar(require("./Task"));
exports.TE = __importStar(require("./TaskEither"));
exports.Sg = __importStar(require("./Semigroup"));
exports.m = __importStar(require("./Monoid"));
exports.Eq = __importStar(require("./Equal"));
exports.Ord = __importStar(require("./Orderable"));
