// @benjstephenson/kittens-ts/dist/src/Task/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sequence = exports.traverse = exports.traverse2 = exports.ap = exports.flatMap = exports.map = exports.of = void 0;
const Task_1 = require("./Task");
const functions_1 = require("../functions");
const of = (a) => new Task_1.Task(() => Promise.resolve(a));
exports.of = of;
const map = (fn, fa) => new Task_1.Task(() => fa.run().then(fn));
exports.map = map;
// of(() => fa.run().then((x) => fn(x)))
const flatMap = (fn, fa) => new Task_1.Task(() => fa.run().then((t) => fn(t).run()));
exports.flatMap = flatMap;
const ap = (fa, fab) => (0, exports.flatMap)((f) => (0, exports.map)(f, fa), fab);
exports.ap = ap;
const traverse2 = (F) => (f, fa) => {
    return undefined;
};
exports.traverse2 = traverse2;
const traverse = (values, f) => new Task_1.Task(() => Promise.all(values.map((x) => f(x).run())));
exports.traverse = traverse;
const sequence = (values) => (0, exports.traverse)(values, functions_1.id);
exports.sequence = sequence;
