// @benjstephenson/kittens-ts/dist/src/Equal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.date = exports.boolean = exports.number = exports.string = exports.record = exports.never = exports.withDefault = exports.from = exports.contravariant = exports.contramap = void 0;
const contramap = (f, eqA) => (0, exports.from)((x, y) => eqA.equals(f(x), f(y)));
exports.contramap = contramap;
exports.contravariant = {
    contramap: exports.contramap,
};
const from = (equals) => ({
    equals,
});
exports.from = from;
const withDefault = () => (0, exports.from)((x, y) => x === y);
exports.withDefault = withDefault;
const never = () => (0, exports.from)((x, y) => false);
exports.never = never;
function record(equalities) {
    return (0, exports.from)((x, y) => {
        return Object.keys(equalities).every((key) => equalities[key].equals(x[key], y[key]));
        // for (const key in equalities) {
        //   if (!equalities[key].equals(x[key], y[key])) {
        //     return false
        //   }
        // }
        // return true
    });
}
exports.record = record;
exports.string = (0, exports.withDefault)();
exports.number = (0, exports.withDefault)();
exports.boolean = (0, exports.withDefault)();
exports.date = (0, exports.contramap)((d) => d.valueOf(), exports.number);
