// @benjstephenson/kittens-ts/dist/src/Orderable/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMonoid = exports.getSemigroup = exports.contravariant = exports.contramap = exports.boolean = exports.string = exports.number = exports.from = void 0;
const from = (compare) => ({
    equals: (x, y) => compare(x, y) === 0,
    compare,
});
exports.from = from;
const primitive = (x, y) => (x < y ? -1 : y > x ? 1 : 0);
exports.number = (0, exports.from)(primitive);
exports.string = (0, exports.from)(primitive);
exports.boolean = (0, exports.from)(primitive);
const contramap = (f, ord) => (0, exports.from)((x, y) => ord.compare(f(x), f(y)));
exports.contramap = contramap;
exports.contravariant = {
    contramap: exports.contramap,
};
const getSemigroup = () => ({
    concat: (x, y) => (0, exports.from)((a, b) => {
        const cmpFirst = x.compare(a, b);
        if (cmpFirst !== 0)
            return cmpFirst;
        else
            return y.compare(a, b);
    }),
});
exports.getSemigroup = getSemigroup;
const getMonoid = () => (Object.assign(Object.assign({}, (0, exports.getSemigroup)()), { empty: (0, exports.from)(() => 0) }));
exports.getMonoid = getMonoid;
