// @benjstephenson/kittens-ts/dist/src/Option/Option.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.None = exports.Some = exports.OptionURI = void 0;
const Eq = __importStar(require("../Equal"));
const E = __importStar(require("../Either"));
const functions_1 = require("./functions");
const instances_1 = require("./instances");
exports.OptionURI = 'Option';
class Some {
    constructor(value) {
        this.value = value;
        this.tag = 'Some';
    }
    equals(other, eq = Eq.withDefault()) {
        return (0, instances_1.getEquals)(eq).equals(this, other);
    }
    isSome() {
        return true;
    }
    isNone() {
        return false;
    }
    get() {
        return this.value;
    }
    map(f) {
        return (0, functions_1.map)(f, this);
    }
    flatMap(f) {
        return (0, functions_1.flatMap)(f, this);
    }
    getOrElse(_default) {
        return this.value;
    }
    getOrCall(_f) {
        return this.value;
    }
    toEither(e) {
        return E.right(this.value);
    }
}
exports.Some = Some;
class None {
    constructor() {
        this.tag = 'None';
    }
    isSome() {
        return false;
    }
    isNone() {
        return true;
    }
    equals(other, eq = Eq.withDefault()) {
        return (0, instances_1.getEquals)(eq).equals(this, other);
    }
    map(f) {
        return (0, functions_1.map)(f, this);
    }
    flatMap(f) {
        return (0, functions_1.flatMap)(f, this);
    }
    getOrElse(fallback) {
        return fallback;
    }
    getOrCall(f) {
        return f();
    }
    toEither(e) {
        return E.left(e);
    }
}
exports.None = None;
