// mismatched/dist/src/matcher/ArrayMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayMatcher = void 0;
const matchMaker_1 = require("../matchMaker/matchMaker");
const MatchResult_1 = require("../MatchResult");
const DiffMatcher_1 = require("./DiffMatcher");
const Mismatched_1 = require("./Mismatched");
const arrayDiff_1 = require("../diff/arrayDiff");
const ofType_1 = require("../ofType");
class ArrayMatcher extends DiffMatcher_1.DiffMatcher {
    constructor(matchers) {
        super();
        this.matchers = matchers;
        this.complexity = DiffMatcher_1.DiffMatcher.andComplexity(matchers);
    }
    static make(expected) {
        return new ArrayMatcher(expected.map(e => matchMaker_1.matchMaker(e)));
    }
    mismatches(context, mismatched, actual) {
        if (!ofType_1.ofType.isArray(actual)) {
            mismatched.push(Mismatched_1.Mismatched.makeExpectedMessage(context, actual, "array expected"));
            return MatchResult_1.MatchResult.wasExpected(actual, this.describe(), 1, 0);
        }
        const pairs = arrayDiff_1.arrayDiff(this.matchers, actual);
        let compares = 0;
        let matches = 0;
        const results = pairs.map((pair) => {
            if (pair.matcher.isSome() && pair.actual.isNone()) {
                compares += 1;
                mismatched.push(Mismatched_1.Mismatched.makeMissing(context, actual, pair.matcher.get().describe()));
                return { [MatchResult_1.MatchResult.expected]: pair.matcher.get().describe() };
            }
            if (pair.matcher.isNone() && pair.actual.isSome()) {
                compares += 1;
                mismatched.push(Mismatched_1.Mismatched.makeUnexpectedMessage(context, actual, pair.actual.get()));
                return { [MatchResult_1.MatchResult.unexpected]: pair.actual.get() };
            }
            if (pair.matcher.isSome() && pair.actual.isSome()) {
                const result = pair.matcher.get().mismatches(context.add("[" + pair.actualIndex + "]"), mismatched, pair.actual.get());
                compares += result.compares;
                matches += result.matchRate * result.compares;
                if (result.passed()) {
                    return pair.actual.get();
                }
                else {
                    return result.diff;
                }
            }
        });
        return new MatchResult_1.MatchResult(results, compares, matches);
    }
    describe() {
        return this.matchers.map(e => e.describe());
    }
}
exports.ArrayMatcher = ArrayMatcher;
