// mismatched/dist/src/matcher/DiffMatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextOfValidationError = exports.DiffMatcher = void 0;
class DiffMatcher {
    constructor() {
        this.complexity = 1;
    }
    static andComplexity(matchers) {
        return matchers.reduceRight((previous, current) => previous + current.complexity, 0);
    }
    static orComplexity(matchers) {
        return matchers.reduceRight((previous, current) => Math.max(previous, current.complexity), 0);
    }
    matches(actual) {
        return this.mismatches(new ContextOfValidationError(), [], actual);
    }
    trialMatches(actual) {
        return this.mismatches(new ContextOfValidationError().inTrial(), [], actual);
    }
}
exports.DiffMatcher = DiffMatcher;
class ContextOfValidationError {
    constructor(context = "actual", isUserDefined) {
        this.context = context;
        this.isUserDefined = isUserDefined;
        this.trialMatch = false;
    }
    add(s) {
        return this.isUserDefined ? this : new ContextOfValidationError(this.context + s).inTrial(this.trialMatch);
    }
    outerContext() {
        return this.isUserDefined ? this.context : '';
    }
    inTrial(trialMatch = true) {
        this.trialMatch = trialMatch;
        return this;
    }
    describe(actual, contextDescription) {
        const outerContext = this.isUserDefined ? this.context : '';
        return new ContextOfValidationError(contextDescription(outerContext, actual), true);
    }
}
exports.ContextOfValidationError = ContextOfValidationError;
